Implement scrollbar-gutter CSS parsing

lukewarlow opened the pull request:
#### 1232ab0a27b75a8941b0e19276d03c4a67748a64
<pre>
Implement scrollbar-gutter CSS parsing
<a href="https://bugs.webkit.org/show_bug.cgi?id=256892">https://bugs.webkit.org/show_bug.cgi?id=256892</a>

Reviewed by Tim Nguyen.

Spec: <a href="https://drafts.csswg.org/css-overflow/#scrollbar-gutter-property">https://drafts.csswg.org/css-overflow/#scrollbar-gutter-property</a>

Grammar for property: auto | stable &amp;&amp; both-edges?

This adds the property to the CSS parser behind the flag, and updates the
relevant expectation files.

It also updates existing tests for scrollbar-gutter to be a bit more
exhaustive.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/inheritance.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/scrollbar-gutter-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/scrollbar-gutter-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/scrollbar-gutter-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/scrollbar-gutter-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForScrollbarGutter):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeScrollbarGutter):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::initialScrollbarGutter):
(WebCore::RenderStyle::scrollbarGutter const):
(WebCore::RenderStyle::setScrollbarGutter):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/ScrollbarGutter.cpp: Added.
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/ScrollbarGutter.h: Added.
(WebCore::operator==):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertScrollbarGutter):

Canonical link: <a href="https://commits.webkit.org/264568@main">https://commits.webkit.org/264568@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cf9ab87a65ec810035eac50d97acdd345fb5fad1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/7984 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8267 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8480 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9635 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8091 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/7988 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10254 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8179 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/10958 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8127 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9215 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7303 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9754 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6516 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7311 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/14890 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7639 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7432 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10791 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/7896 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6422 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7211 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/7224 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1912 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11419 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7637 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/animation/CSSPropertyAnimation.cpp:
```diff
@@ -3699,6 +3700,7 @@ CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap()
         new DiscreteSVGPropertyWrapper<const String&>(CSSPropertyMarkerEnd, &SVGRenderStyle::markerEndResource, &SVGRenderStyle::setMarkerEndResource),
         new DiscreteSVGPropertyWrapper<const String&>(CSSPropertyMarkerMid, &SVGRenderStyle::markerMidResource, &SVGRenderStyle::setMarkerMidResource),
         new DiscreteSVGPropertyWrapper<const String&>(CSSPropertyMarkerStart, &SVGRenderStyle::markerStartResource, &SVGRenderStyle::setMarkerStartResource),
+        new DiscretePropertyWrapper<ScrollbarGutter>(CSSPropertyScrollbarGutter, &RenderStyle::scrollbarGutter, &RenderStyle::setScrollbarGutter),
```
Work out why this fails to compile

lukewarlow commented on LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
```diff
@@ -1172,6 +1172,12 @@ const gCSSProperties2 = {
+    types: [
+      { type: 'discrete', options: [ [ 'auto', 'stable' ], [ 'auto', 'stable both-edges' ], [ 'stable', 'stable both-edges' ] ] }
+    ]
+  },
```
https://github.com/web-platform-tests/wpt/pull/40167

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1214,6 +1214,14 @@ static Ref<CSSValueList> valueForScrollSnapAlignment(const ScrollSnapAlign& alig
+    if (!gutter.bothEdges)
+        return CSSValueList::createSpaceSeparated(createConvertingToCSSValueID(gutter.type));
+    return CSSValueList::createSpaceSeparated(createConvertingToCSSValueID(gutter.type),
+        CSSPrimitiveValue::create(CSSValueBothEdges));
```
You can probably just use CSSValuePair

nt1m commented on Source/WebCore/css/ComputedStyleExtractor.cpp:
```diff
@@ -1214,6 +1214,14 @@ static Ref<CSSValueList> valueForScrollSnapAlignment(const ScrollSnapAlign& alig
+static Ref<CSSValueList> valueForScrollbarGutter(const ScrollbarGutter& gutter)
+{
+    if (!gutter.bothEdges)
+        return CSSValueList::createSpaceSeparated(createConvertingToCSSValueID(gutter.type));
```
This can just be a single CSSPrimitiveValue

nt1m commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -6669,6 +6669,30 @@ RefPtr<CSSValue> consumeScrollSnapType(CSSParserTokenRange& range)
+        return CSSValueList::createSpaceSeparated(secondValue.releaseNonNull(), firstValue.releaseNonNull());
+
+    // Only valid cobination of two values is stable && both-edges or both-edges && stable
+    return nullptr;
```
I would have written it like this:

```suggestion
    if (auto ident = consumeIdent<CSSValueAuto>(range))
        return CSSPrimitiveValue::create(CSSValueAuto);
    
    if (auto first = consumeIdent<CSSValueStable>(range)) {
        if (auto second = consumeIdent<CSSValueBothEdges>(range))
            return CSSValuePair::create(...);
        return first;
    }
 
    if (auto first = consumeIdent<CSSValueBothEdges>(range)) {
        if (auto second = consumeIdent<CSSValueStable>(range))
            return CSSValuePair::create(...);
        return nullptr;
    }
    
    return nullptr;
```

nt1m commented on Source/WebCore/rendering/style/StyleScrollbarGutter.h:
```diff
@@ -0,0 +1,48 @@
+struct ScrollbarGutter {
+    ScrollbarGutterType type { ScrollbarGutterType::Auto };
+    bool bothEdges { false };
+};
```
I think two booleans would be enough to represent this (the code around it would need to be adjusted):

```suggestion
struct ScrollbarGutter {
    bool isAuto { true };
    bool bothEdges { false };
};
```

nt1m commented on Source/WebCore/rendering/style/StyleScrollbarGutter.cpp:
Can we name this ScrollbarGutter.cpp & ScrollbarGutter.h ?

nt1m commented on LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/scrollbar-gutter-invalid-expected.txt:
```diff
@@ -13,4 +13,11 @@ PASS e.style['scrollbar-gutter'] = "3em" should not set the property value
+PASS e.style['scrollbar-gutter'] = "always" should not set the property value
+PASS e.style['scrollbar-gutter'] = "always both" should not set the property value
+PASS e.style['scrollbar-gutter'] = "always force" should not set the property value
+PASS e.style['scrollbar-gutter'] = "always both force" should not set the property value
```
It would be interesting to test some bad combinations with `both-edges`

lukewarlow replied:
Should I do that as a PR in the WPT repo or here?

nt1m replied:
Here, and then exported to WPT would be ideal

lukewarlow replied:
Have added those.

When I try to run the export script I get this

`python3 Tools/Scripts/export-w3c-test-changes`

"No changes to upstream. Exiting..."

nt1m replied:
`export-w3c-test-changes -g HEAD -b 256892 --no-linter -c`

lukewarlow replied:
https://github.com/web-platform-tests/wpt/pull/40250

lukewarlow committed:
Implement scrollbar-gutter CSS parsing
https://bugs.webkit.org/show_bug.cgi?id=256892

Reviewed by Tim Nguyen.

Spec: https://drafts.csswg.org/css-overflow/#scrollbar-gutter-property

Grammar for property: auto | stable && both-edges?

This adds the property to the CSS parser behind the flag, and updates the
relevant expectation files.

It also updates existing tests for scrollbar-gutter to be a bit more
exhaustive.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/inheritance-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/inheritance.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/scrollbar-gutter-invalid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/scrollbar-gutter-invalid.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/scrollbar-gutter-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/scrollbar-gutter-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/property-list.js:
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ios-wk2/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/ipad/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/css/css-cascade/all-prop-revert-layer-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/accumulation-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/addition-per-property-002-expected.txt:
* LayoutTests/platform/mac-wk1/imported/w3c/web-platform-tests/web-animations/animation-model/animation-types/interpolation-per-property-002-expected.txt:
* LayoutTests/platform/wpe/imported/w3c/web-platform-tests/css/css-cascade/all-prop-initial-xml-expected.txt:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::valueForScrollbarGutter):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::initialValueForLonghand):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeScrollbarGutter):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::initialScrollbarGutter):
(WebCore::RenderStyle::scrollbarGutter const):
(WebCore::RenderStyle::setScrollbarGutter):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/ScrollbarGutter.cpp: Added.
(WebCore::operator<<):
* Source/WebCore/rendering/style/ScrollbarGutter.h: Added.
(WebCore::operator==):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.cpp:
(WebCore::StyleRareNonInheritedData::StyleRareNonInheritedData):
(WebCore::StyleRareNonInheritedData::operator== const):
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/style/StyleBuilderConverter.h:
(WebCore::Style::BuilderConverter::convertScrollbarGutter):

Canonical link: https://commits.webkit.org/264568@main
